src/Features/Core/Portable/Diagnostics/DiagnosticService.cs

```diff
@@ -337,16 +363,29 @@ private struct Data : IEquatable<Data>
 
             public Data(UpdatedEventArgs args, ImmutableArray<DiagnosticData> diagnostics)
             {
-                this.Workspace = args.Workspace;
+                // most of data has same workspace, so this should reduce us creating new
+                // weak reference for same workspace multiple times.
+                _workspace = s_cache.GetValue(args.Workspace, w => new WeakReference<Workspace>(w));
+
                 this.ProjectId = args.ProjectId;
                 this.DocumentId = args.DocumentId;
                 this.Id = args.Id;
                 this.Diagnostics = diagnostics;
             }
 
+            public Workspace Workspace => _workspace.GetTarget();
+
             public bool Equals(Data other)
             {
-                return this.Workspace == other.Workspace &&
+                var workspace1 = Workspace;
+                var workspace2 = other.Workspace;
+
+                if (workspace1 == null || workspace2 == null)
+                {
```

Comment:
Why is the preview workspace not cleaning diagnostics it reported?  How can we make it so that no workspaces can be created that do not clean up after themselves when they do away.